group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicMetadata.java

```diff
@@ -41,10 +41,25 @@ public class TopicMetadata {
      */
     private final int numPartitions;
 
+    /**
+     * Map of every partition Id to a set of its rack Ids, if they exist.
+     * If rack information is unavailable for all partitions, this is an empty map.
+     */
+    private final Map<Integer, Set<String>> partitionRacks;
```

Comment:
Maybe we should revert the structure from `partition -> racks` to `rack -> partitions` for the following reasons:

1. The number of racks is typically smaller than the number of partitions, so this new structure can reduce duplicate information.
2. The size of a rack (string) is usually larger than a partition (int), so it's better to duplicate partitions rather than racks.
3. If the size of the new structure is acceptable, we could use it instead of a hash to avoid calculating a hash for the entire structure, as we can leverage a map/set for a quick equality check.

Reply:
Hi @chia7712, thanks for the suggestion. With this way, we still need to store `subscriptionMetadata` in `__consumer_offsets`, or we can't get an old value to compare with new calculated value.

I have a rough idea. How about we always calculate new target assignment when the metadata is expired? If latest target assignment is not equal to the old one, we bump group epoch. With this way, we can avoid bumping useless group epoch. Also, we don't store `subscriptionMetadata` in records.

Reply:
> How about we always calculate new target assignment when the metadata is expired? If latest target assignment is not equal to the old one, we bump group epoch. 

Does `PartitionAssignor` guarantee idempotency ? For example, if metadata expires due to an unrelated topic change, we might end up with a new assignment that's not equal to the old one, causing an unnecessary rebalance.

Reply:
> Does `PartitionAssignor` guarantee idempotency ?

From source code, I think no. The `PartitionAssignor#assign` takes `GroupSpec`. The `GroupSpec#memberSubscription` returns member data. The `MemberSubscription#subscribedTopicIds` only returns subscribed topic ids. In conclusion, `PartitionAssignor` can only get subscribed topics from `GroupSpec`, so unrelated topics don't change the result.

https://github.com/apache/kafka/blob/d7ac865fb0d2f739392623a379d056a7bf3c19f4/group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/PartitionAssignor.java#L34-L45

https://github.com/apache/kafka/blob/d7ac865fb0d2f739392623a379d056a7bf3c19f4/group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/GroupSpec.java#L45-L52

https://github.com/apache/kafka/blob/d7ac865fb0d2f739392623a379d056a7bf3c19f4/group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/MemberSubscription.java#L44-L50

Reply:
> In conclusion, PartitionAssignor can only get subscribed topics from GroupSpec, so unrelated topics don't change the result.

my point was the approach you mentioned will trigger the assignment even though the metadata is changed by unrelated topic. If assignor is not idempotent, the new assignment will be not equal to old one. That means "unrelated topic" can will trigger "unnecessary" rebalance.

Reply:
According to @dajac's comment (https://github.com/apache/kafka/pull/17444#discussion_r1816397624), it seems the assignor should generate no-op assignments for consumers when the subscribed topics haven't changed. However, I'm not sure if all assignors are expected to produce deterministic results. I know many balancing algorithms rely on random search, which means they may not generate the same assignment even with the same subscribed topics.

Reply:
Hi @chia7712, thanks for the comments. Ideally, assignors should generate deterministic results, so it doesn't trigger useless rebalance.

I have another approach to determine when do we need to generate a new target assignment if there is new metadata image.

In `GroupMetadataManager#onNewMetadataImage`, it gets new metadata image and metadata delta, so we can know which topics are updated in the new metadata. Also, we have a `GroupMetadataManager#groupsByTopics` data, so we know which group needs to rebalance when there is topic update.

For metadata change, there are two cases can trigger group rebalance:
1. There is new partition in a topic.
2. Rack information is changed. For `broker.rack` change, it can only happen when we restart a broker.

In `TopicDelta`, we can get new partition by `TopicDelta#newPartitions`, so this solves the first case.
In `PartitionRegistration`, we can get `removingReplicas` and `addingReplicas`. The partition racks is calculated by partition replicas. The `broker.rack` can only be changed when broker restarts. That means when broker stop/start, it's in `removingReplicas`/`addingReplicas`.

Finally, we can set a new value in `ModernGroup`. When above cases are detected in `GroupMetadataManager#onNewMetadataImage`, we set the value to true. In new consumer group heartbeat, we don't need to calculate subscription metadata, we can check the value and calculate a new target assignment directly. After we gets new target assignment, we set the value back to false.

Reply:
>  Ideally, assignors should generate deterministic results, so it doesn't trigger useless rebalance.

I don't think this is easy to follow. For instance, assignors typically loop through members, topics, or partitions to calculate the assignment, but we don't enforce a specific order for these objects in the input.

> For metadata change, there are two cases can trigger group rebalance:

The rack of a partition can change if the partition is moved from node_a in rack_a to node_b in rack_b.

> Finally, we can set a new value in ModernGroup. When above cases are detected in GroupMetadataManager#onNewMetadataImage, we set the value to true. In new consumer group heartbeat, we don't need to calculate subscription metadata, we can check the value and calculate a new target assignment directly. After we gets new target assignment, we set the value back to false.

Reply:
@FrankYang0529 Sorry for opening too many discussions here. As @dajac suggested, it might be good to create a KIP. You could summarize the four possible solutions and move the discussion to the dev channel. 

There are two approaches from the JIRA: one from me (optimizing the structure), and yours (leveraging topic delta).

Reply:
> I don't think this is easy to follow. For instance, assignors typically loop through members, topics, or partitions to calculate the assignment, but we don't enforce a specific order for these objects in the input.

@chia7712 All the assignors are sticky and all new assignors must be sticky too. This is a fundamental principle in KIP-848. That being said, as you said, it depends on the implementation of the assignor so one can ignore this. Hence they would get what they expect.

> We still need to track the 'state' of the current assignment to determine whether 'the change is worth triggering a rebalance,' correct? I mean, how do we ensure that the changes from GroupMetadataManager#onNewMetadataImage actually require a rebalance?

We tack the topology of the group. This includes two things: 1) the members and 2) the subscribed topics' partitions metadata. We trigger a rebalance whenever on of them changes. Here we discuss how we could extend 2) to also cover racks.

> my point was the approach you mentioned will trigger the assignment even though the metadata is changed by unrelated topic.

This is incorrect. See my previous point. We actually have 2) to only trigger a rebalance when the subscribed topics have changed.

Reply:
> Maybe we should revert the structure from `partition -> racks` to `rack -> partitions` for the following reasons:
> 
> 1. The number of racks is typically smaller than the number of partitions, so this new structure can reduce duplicate information.
> 2. The size of a rack (string) is usually larger than a partition (int), so it's better to duplicate partitions rather than racks.
> 3. If the size of the new structure is acceptable, we could use it instead of a hash to avoid calculating a hash for the entire structure, as we can leverage a map/set for a quick equality check.

I'd like to get rid of `TopicMetadata` :).

Reply:
> All the assignors are sticky and all new assignors must be sticky too. This is a fundamental principle in KIP-848. That being said, as you said, it depends on the implementation of the assignor so one can ignore this. Hence they would get what they expect.

Do you mean all assignors should always generate the identical assignment for the same input? If so, do we still trace the topic metadata for each group? We can always try to call assignor to generate the new assignment when the related topics get changed by GroupMetadataManager#onNewMetadataImage. This approach is equal to @FrankYang0529 comment (https://github.com/apache/kafka/pull/17444#discussion_r1817615221). Additionally, we can check the assignment change to determine  whether the group epoch bump is necessary.

> This is incorrect. See my previous point. We actually have 2) to only trigger a rebalance when the subscribed topics have changed.

Sorry for the unclear comment. What I meant was that unrelated topic changes, such as follower movements, can trigger a rebalance if we don’t track the topic metadata (similar to my comment before https://github.com/apache/kafka/pull/17444#discussion_r1817644544)

---

In short, the coordinator can perform fewer checks for topic changes if we trust that the assignor is sticky and can minimize partition movements.

Reply:
Hi @chia7712 / @dajac, thanks for the great discussion here. I write a draft [KIP-1101](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1101%3A+Trigger+rebalance+on+rack+topology+changes) about using `TopicDelta` to detect changes (https://github.com/apache/kafka/pull/17444#discussion_r1817877601). I will finish test plan and rejective alternatives tomorrow. Before creating a discussion email thread, I would like to check do we on the same page about this method. Thanks!

Reply:
@chia7712 This is not how we have designed it. The group epoch is bumped to trigger a re-computation of the target assignment. If the assignment does not change, only the epoch of the target assignment is bump. End of the operation. We don't need to compare assignments to not bump the group epoch is the first place.

Reply:
@FrankYang0529 Your proposal does not work because the changes may happen when the group coordinator is offline. This is the reason why we have the current mechanism in place. I think that we need to keep something similar. Another thing is that we want to based this on the replicas set, not the ISR. When a broker shutdowns, it remains in the replicas set. The adding/removing replicas sets are only used when there is a reassignment.

Reply:
@dajac it seems the idea approach needs following advantages:

1. less computing
2. less stored objects
3. less new assignment
4. less coupling 

Maybe we can leverage [MetadataProvenance](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/image/MetadataProvenance.java)? It has an incremented offset that changes with new metadata images and is small enough. We could store the offset in the group coordinator when the topic topology changes. This approach addresses 'fewer stored objects,' 'fewer new assignments,' and 'less coupling,' but still requires some computation to check topology changes.